sklearn/conftest.py

```diff
@@ -245,6 +245,113 @@ def pyplot():
     pyplot.close("all")
 
 
+def is_xdist_controller(config):
+    return not hasattr(config, "workerinput")
+
+
+def pytest_generate_tests(metafunc):
+    """Parametrization of global_random_seed fixture
+
+    based on the SKLEARN_TESTS_GLOBAL_RANDOM_SEED environment variable.
+
+    The goal of this fixture is to prevent tests that use it to be sensitive
+    to a specific seed value while still being deterministic by default.
+
+    See the documentation for the SKLEARN_TESTS_GLOBAL_RANDOM_SEED
+    variable for instructions on how to use this fixture.
+
+    https://scikit-learn.org/dev/computing/parallelism.html#sklearn-tests-global-random-seed
+
+    """
+    # metafunc.config["random_seeds"] has been set in the
+    # handle_global_random_seed function based on
+    # SKLEARN_TESTS_GLOBAL_RANDOM_SEED
+    random_seeds = metafunc.config.getoption("random_seeds")
+
+    if "global_random_seed" in metafunc.fixturenames:
+        metafunc.parametrize("global_random_seed", random_seeds)
+
+
+def handle_global_random_seed(config):
+    """Part of pytest_configure that deals with the global_random_seed fixture."""
+    RANDOM_SEED_RANGE = list(range(100))  # All seeds in [0, 99] should be valid.
+    random_seed_var = environ.get("SKLEARN_TESTS_GLOBAL_RANDOM_SEED")
+
+    default_random_seeds = [42]
+
+    # When using pytest-xdist this function is called both in the xdist
+    # controller and xdist workers. Care is needed in the
+    # SKLEARN_TESTS_GLOBAL_RANDOM_SEED == 'any' case to generate a random seed
+    # in the xdist controller and reuse it in the xdist workers.
+    if random_seed_var == "any":
+        # inside the xdist controller: pick-up one seed at random in the range
+        # of admissible random seeds.
+        if is_xdist_controller(config):
+            random_seeds = [Random().choice(RANDOM_SEED_RANGE)]
+        else:
+            # inside a xdist worker: reuse random_seeds that have been
+            # generated in the xdist controller and passed to the xdist workers
+            # by pytest_configure_node. For some unknown reason, pytest_configure_node
+            # is not executed in some edge cases, for example:
+            # pytest -n2 --pyargs sklearn.tests.test_dummy
+            # In these edge cases, random_seeds is set to a fixed value
+            random_seeds = getattr(
+                config.workerinput, "random_seeds", default_random_seeds
+            )
```

Comment:
It does not seem to pickup the seed randomly selected by the controller:

```
SKLEARN_TESTS_GLOBAL_RANDOM_SEED="any" pytest -n 8 -v sklearn/cluster/tests/test_k_means.py -k test_sample_weight_init
====================================================================================================== test session starts =======================================================================================================
platform darwin -- Python 3.11.6, pytest-7.4.2, pluggy-1.3.0 -- /Users/ogrisel/miniforge3/envs/dev/bin/python3.11
cachedir: .pytest_cache
To reproduce this test run, set the following environment variable:
    SKLEARN_TESTS_GLOBAL_RANDOM_SEED="89"
See: https://scikit-learn.org/dev/computing/parallelism.html#sklearn-tests-global-random-seed
rootdir: /Users/ogrisel/code/scikit-learn
configfile: setup.cfg
plugins: repeat-0.9.2, anyio-4.0.0, cov-4.1.0, xdist-3.3.1
8 workers [2 items]     
scheduling tests via LoadScheduling

sklearn/cluster/tests/test_k_means.py::test_sample_weight_init[42-random] 
sklearn/cluster/tests/test_k_means.py::test_sample_weight_init[42-k-means++] 
[gw0] [ 50%] PASSED sklearn/cluster/tests/test_k_means.py::test_sample_weight_init[42-random] 
[gw1] [100%] PASSED sklearn/cluster/tests/test_k_means.py::test_sample_weight_init[42-k-means++] 

======================================================================================================= 2 passed in 1.69s ========================================================================================================
```

It's always running the tests with 42 while it mentions 89 in the test report header message.

Reply:
It does work as expected when I do not enable xdist though:

```
====================================================================================================== test session starts =======================================================================================================
platform darwin -- Python 3.11.6, pytest-7.4.2, pluggy-1.3.0 -- /Users/ogrisel/miniforge3/envs/dev/bin/python3.11
cachedir: .pytest_cache
To reproduce this test run, set the following environment variable:
    SKLEARN_TESTS_GLOBAL_RANDOM_SEED="82"
See: https://scikit-learn.org/dev/computing/parallelism.html#sklearn-tests-global-random-seed
rootdir: /Users/ogrisel/code/scikit-learn
configfile: setup.cfg
plugins: repeat-0.9.2, anyio-4.0.0, cov-4.1.0, xdist-3.3.1
collected 280 items / 278 deselected / 2 selected                                                                                                                                                                                

sklearn/cluster/tests/test_k_means.py::test_sample_weight_init[82-k-means++] PASSED                                                                                                                                        [ 50%]
sklearn/cluster/tests/test_k_means.py::test_sample_weight_init[82-random] PASSED                                                                                                                                           [100%]

=============================================================================================== 2 passed, 278 deselected in 0.17s ================================================================================================
```

Reply:
Hum this is what you commented above:

> as I mentioned in one of the comment pytest --pyargs sklearn.tests.test_dummy -n2 does not seem to run pytest_configure_node so in the 'any' case we set a fixed random_state (the default value). I don't think this matters that much in practice.

I don't understand why you think it does not matter. The "any" mode is used by our nightly CI to discover unexpected seed-sensitive test failures over time. I think it's important.

As I commented above this is not related to the use of `--pyargs`: I do not use it in the reproducer.

Reply:
Thanks for testing, I thought I tested a variety of combinations and only the `--pyargs sklearn.<some-subpackage>` did not work, which is why I said it is not that important. Apparently it is more complicated than this ...

I need to check again!

Reply:
So I fixed the bug I think. I was using `getattr(config.workerinput, ...)` rather than `config.workerinput.get` ...

I tested the cases mentioned in my previous comment and now only 4. does not work.

Reply:
I believe the issue is actually https://github.com/pytest-dev/pytest-xdist/issues/917. `pytest_configure` is not run in the xdist controller in 4. for some reason so the mechanism to create the random seed in the main controller and access it in the xdist workers does not work.

Reply:
Would it be possible to generate the seed in each worker instead of having to use the controller?

I am thinking of something like calling the random number generator with a shared seed in each worker. To have a shared seed we need some form of state that is easily accessed by each worker. I don't know if such a thing exists in pytest, maybe each run is assigned some kind of "run ID" or something?

If each worker can derive the correct seed itself we don't need to have to use the controller to coordinate the workers.

Reply:
Maybe https://pytest-xdist.readthedocs.io/en/stable/how-to.html#uniquely-identifying-the-current-test-run?

Reply:
Interesting idea, I'll look into it!

Reply:
Thinking about it, you actually want the controller to know about the seed, be it only to display it in `pytest_report_header`.

I looked a bit at `PYTEST_XDIST_TESTRUNUID` env variable but it is only available in the workers and not the controller, there is the `testrun_uid` fixture but I am not sureit can be made to fit nicely with  the `pytest_generate_tests` with `metafunc` thing.